source_code/main_mcu/src/LOGIC/logic_totp.c

```diff
@@ -0,0 +1,174 @@
+                    // index remaining, that have to be put in the next byte. To get the number places the remaining bits have to be shifted in the next byte, we have to subtract
+                    // shift_offset from the number of bits in a byte. If we rearrange the order in that term, we get shift_offset = shift_offset + 8 - 5 = shift_offset + 3)
+                    // As an example for the second character, shift_offset is 3. So we take the first 3 of the 5 index bits and put them into current byte. Then we have 2 bits remaining.
+                    // Those bits need to be shifted then by 6 places to the left, to be at the beginning of the next byte.
+                    decoded_key[decoded_index] |= j >> (LOGIC_TOTP_BASE32_BITS - shift_offset);
+                    shift_offset += LOGIC_TOTP_BITS_PER_BYTE - LOGIC_TOTP_BASE32_BITS;
+                    decoded_index++;
+                    decoded_key[decoded_index] |= j << (shift_offset);
+                }
+            }
+        }
+    }
+}
+
+
+/*! \fn     logic_totp(char* totp_token, const char* key)
+*   \brief  Calculate TOTP Token using given key
+*   \param  calculated totp_token, char[6] expected
+*   \param  Base32 encoded key that is used to calculate totp_token, char[16] expectd
+*   \return void
+*/
+void logic_totp(char* totp_token, const char* key)
+{
+    char hash[20];
+    char decoded_key[10];
+    int truncated_hash = 0;
+    int offset = 0;
+    int shift_offset = LOGIC_TOTP_BITS_PER_BYTE * 3;
+    int token_idx = LOGIC_TOTP_TOKEN_LEN - 1;
+    time_t timestamp = logic_totp_get_time();
+    // step is the message the hash gets calculated for. TOTP is based on the current unix timestamp. As a TOTP token is valid for 30 seconds, we divide the current timestamp by
+    // 30 before calculating the SHA1 HMAC hash
+    time_t step = utils_switch_endiannes(timestamp/30);
+
+    logic_totp_base32_decode_key(decoded_key, key);
+    logic_totp_sha1_hmac(hash, decoded_key, &step);
+
+    // We use the last nibble of the calculated hash to get the offset. This is the offset in the hash, from where we take 4 bytes as truncated hash.
+    offset = hash[19] & 0x0F;
+
+    for (int i = offset; i < offset + LOGIC_TOTP_TRUNCATED_HASH_LEN; i++)
+    {
+        // Get the truncated hash as int, as we those 4 bytes as a whole number.
+        truncated_hash |= (hash[i] << shift_offset) & (0x000000FF << shift_offset);
+        shift_offset -= LOGIC_TOTP_BITS_PER_BYTE;
+    }
+
+    // Initialize the totp_token with 0 characters. This is used for leading zeros, if the truncated hash is less than 100000.
+    memset(totp_token, '0', LOGIC_TOTP_TOKEN_LEN);
+
+    // Usually you would now first calculate truncated_hash mod 1000000 to have only 6 places remaining. We don't need that, as we take a maximum of 6 places from truncated_hash,
```

Comment:
I think you mean "digits" and not places?
I don't follow your reasoning... could you explain to me again
The truncated hash is an uint32_t, between 0 and 4,294,967,295 and the spec does specify the use of %

Reply:
The idea was, that we don't need to calulate mod 1000000 to get the last 6 digits of the truncated hash, and then convert into a char array, because the termination condition of that for loop would guarantee, that we only get maximum 6 digits from truncated hash. But as this might be confusing, I will add that

Reply:
I see.... I wonder what's faster, mod or /=

Reply:
Cleaner in my opinion to just use built-ins. We don't really need speed here. Its not like this is happening at 1000Hz...Why not keep it simple (KISS)? Two lines: hash % 1000000 and sprintf("%06u" ...)?

Also, similar to the question about the time step? Do we want to support more than 6 digits as configurable?
From RFC4226:
Implementations MUST extract a 6-digit code at a minimum and possibly 7 and 8-digit code.  Depending on security requirements, Digit = 7 or more SHOULD be considered in order to extract a longer HOTP value.

Reply:
in the database node we could have a field that specifies that: # of digits, key size...

Reply:
Same answer as below. If we support different token lengths, do you now any service, that needs 7 oder more digits? And how is that transferred? The regular way to add TOTP Services is to enter manually the secret (base32 encoded) or scan a barcode, that also only contains the secret. Haven't seen a different method so far.

Reply:
that's a good point... Maybe services discover the TOTP size the moment the first code is entered?